RHDEVDOCS-3878 - 5.4 Release Notes (#44624)
openshift/openshift-docs: modules/cluster-logging-loki-tech-preview.adoc

// Module included in the following assemblies:
//cluster-logging-release-notes.adoc

:_content-type: PROCEDURE
[id="cluster-logging-loki-tech-preview_{context}"]
:FeatureName: Loki Operator
include::snippets/technology-preview.adoc[]

> [id="cluster-logging-about-loki"]

Reviewer: I missing the following prerequisites for Loki-Operator:
```
.Prerequisites

* {product-title}: 4.10
* {logging-title-uc}: 5.4
 ```

> = About the Loki Operator

Author: Looks like ASCIIDoc isn't including the module correctly for some reason.

> = About the Loki Operator

Author: @kabirbhartiRH - should be corrected.

[role="_abstract"]
Loki is a horizontally scalable, highly available, multi-tenant log aggregation system currently offered as an alternative to Elasticsearch as a log store for the {logging}.

> == Deploying the Lokistack

Reviewer 2: I only see one instance of "Lokistack" in this PR. Do you mean "Loki stack" or "LokiOperator"?

Author: Loki Operator is the Operator name. 
LokiStack is the schema for the lokistacks API. 
Loki is the component name / upstream project's name.

> You can use the {product-title} web console to install the LokiOperator.

Reviewer 3: These two lines are appearing on the same line on the preview docs.

Reviewer 2: LokiOperator seems strange.  For the first instance, at least, I'd expect something like this: 
```suggestion
You can use the {product-title} web console to install the Loki Operator named "LokiOperator."
```

Reviewer 3: Resolved. looks good to now

To install the LokiOperator using the {product-title} web console:

. Install the LokiOperator:

.. In the {product-title} web console, click *Operators* -> *OperatorHub*.

.. Choose  *LokiOperator* from the list of available Operators, and click *Install*.

> .. Ensure that the *All namespaces on the cluster* is selected under *Installation Mode*.

Reviewer 2: ```suggestion
.. Ensure that *All namespaces on the cluster* is selected under *Installation Mode*.
```
Or maybe
```suggestion
.. Under *Installation Mode*, select *All namespaces on the cluster*.
```

.. Ensure that *openshift-operators-redhat* is selected under *Installed Namespace*.
+
You must specify the `openshift-operators-redhat` namespace. The `openshift-operators`
namespace might contain Community Operators, which are untrusted and could publish
a metric with the same name as an {product-title} metric, which would cause
conflicts.

.. Select *Enable operator recommended cluster monitoring on this namespace*.
+
This option sets the `openshift.io/cluster-monitoring: "true"` label in the Namespace object.
You must select this option to ensure that cluster monitoring
scrapes the `openshift-operators-redhat` namespace.

.. Select an *Approval Strategy*.
+
* The *Automatic* strategy allows Operator Lifecycle Manager (OLM) to automatically update the Operator when a new version is available.
+
* The *Manual* strategy requires a user with appropriate credentials to approve the Operator update.

.. Click *Install*.

> .. Verify that the LokiOperator installed by switching to the *Operators* → *Installed Operators* page.

Reviewer 2: ```suggestion
.. Verify that you installed the LokiOperator. Visit the *Operators* → *Installed Operators* page and look for "LokiOperator." 
```

.. Ensure that *LokiOperator* is listed in all the projects whose *Status* is *Succeeded*.

> [NOTE]

Reviewer 4: How about drop this ClusterLogForwarder sentence. I think this the sentence in vector section.

> .. Ensure that *LokiOperator* is listed in all projects with a *Status* of *Succeeded*.

Reviewer 2: ```suggestion
.. Ensure that *LokiOperator* is listed in all the projects whose *Status* is *Succeeded*.
```

====
You do not have to specify a secret in the `ClusterLogForwarder` CR if Vector is forwarding to Loki using token authentication.
====

[role="_additional-resources"]
.Additional resources

> * link:https://grafana.com/docs/loki/latest/[Loki Documentation]

Reviewer 4: The Loki Documentation doesn't include anything about how to deploy lokistack via Operator.   it is better to link the PR If had created.  or else https://github.com/grafana/loki/tree/main/operator/docs

Reviewer: Good catch @anpingli (sorry for being late on reviewing the docs). But I suggest to add some simple references as:
1. To install a simple lokistack use [manifest](https://github.com/grafana/loki/blob/main/operator/hack/lokistack_gateway_ocp.yaml)
2. To forward logs from fluentd/vector on OpenShift link: https://github.com/grafana/loki/blob/main/operator/docs/forwarding_logs_to_gateway.md#openshift-logging

For 1. above we need some docs to:
-  Explain users how to create and reference on the object storage secret (e.g. for [AWS S3](https://github.com/grafana/loki/blob/main/operator/hack/deploy-aws-storage-secret.sh))
- As well a small table explaining the CPU/Mem and PVC requirements when selecting one of the three sizes `1x.extra-small`, `1x.small`, 1x.medium` as below:

| Size | Total CPU required | Total Mem required | Total PVC required | Description |
| ------ | ------------------------: | -------------------------: | ------------------------: | --------------- |
| 1x.extra-small | TBD | TBD | TBD | This size provides an non-HA Loki installation for demo-purposed only. DO NOT USE IN PRODUCTION |
| 1x.small |  TBD | TBD | TBD | This size provide an HA Loki installation with single replication factor. Means all chunks store operations are replicated to two ingesters at a time. The Log chunks are stored once on object storage, means only the succeeding ingester wins the race. It targets Log Volume ingestion of 250GB/day and Query Traffic of 50 req/s |
| 1x.small |  TBD | TBD | TBD | This size provide an HA Loki installation with double replication factor. Means all chunks store operations are replicated to three ingesters at a time. The Log chunks are stored once on object storage, means only the succeeding ingester wins the race. It targets Log Volume ingestion of 2TB/day and Query Trafic of 75 req/s |

I will post here the TBD for CPU/Mem/PVC later.

Author: > The Loki Documentation doesn't include anything about deploy lokistack via Operator. it is better to link the PR If had created or else link these https://github.com/grafana/loki/tree/main/operator/docs

Was more included for general Loki additional info -- moved up to the bottom of the about section.

Author: @periklis / @anpingli - Additional docs for Loki are covered by RHDEVDOCS-3138, we wouldn't include everything about Loki in the Release Notes... just, this exists, here's how to turn it on.

Reviewer: Fine for me if we are tracking this in a follow up ticket

// Module included in the following assemblies:
//cluster-logging-release-notes.adoc

:_content-type: PROCEDURE
[id="cluster-logging-loki-tech-preview_{context}"]
:FeatureName: Loki Operator
include::snippets/technology-preview.adoc[]

[id="cluster-logging-about-loki"]

> = About the Loki Operator

Reviewer 3: This is currently not displayed as a headline

[role="_abstract"]

> Loki is currently offered as an alternative to Elasticsearch as a log store for the {logging}.

Reviewer 3: This could be a bit better? Loki is a horizontally scalable, highly available, multi-tenant log aggregation system and is currently planned to be offered as an alternative to Elasticsearch as a log store for the logging subsystem

WDYT?

Reviewer 3: Gotcha. I mentioned that considering we still haven't GA'd it yet.

> = Deploying the Loki Collector

Reviewer 4: @libander  Loki is not a collector.      We should use loki-operator, lokistack.

.Deploy loki-operator under openshift-operators namespace
[source, terminal]
----

> oc create -f  aosqe-tools/logging/log_template/loki-operator/loki-og-sub.yaml

Reviewer 4: It is better to guide user to deploy loki-operator using Webconsole.
To deploy Loki Operator via client.

- Deploy openshift-operators-redhat namespace if it doesn't exist
cat <<EOF | oc create -f -
kind: Namespace
apiVersion: v1
metadata:
  name: openshift-operators-redhat
  annotations:
    openshift.io/node-selector: ""
  labels:
    openshift.io/cluster-logging: "true"
    openshift.io/cluster-monitoring: "true"
EOF

- Deploy OperatorGroup under openshift-operators-redhat if no OperatorGroup exists

cat <<EOF |oc create -f -
apiVersion: operators.coreos.com/v1
kind: OperatorGroup
metadata:
  name: openshift-operators-redhat
  namespace: openshift-operators-redhat
spec: {}
EOF

- subscript to CatalogSource redhat-operators  to deploy loki-operator

cat <<EOF |oc create -f -
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: loki-operator
  namespace: openshift-operators-redhat
spec:
  channel: "candidate"
  installPlanApproval: Automatic
  name: loki-operator
  source: redhat-operators
  sourceNamespace: openshift-marketplace
EOF

----

.Set Environment variables
[source, yaml]
----
export NAMESPACE=openshift-logging
export SECRETNAME=credsecret
export AWS_DEFAULT_REGION=us-east-2
export LOKI_BUCKET_NAME="aosqe-loki-bucket"
export ENDPOINT="https://s3.${REGION}.amazonaws.com"
export AWS_ACCESS_KEY_ID=<AWS_ACCESS_KEY_ID>
export AWS_SECRET_ACCESS_KEY=<AWS_SECRET_ACCESS_KEY>
export STORAGE_CLASS=<storage_class>

> export LOKI_BUCKET_NAME=loki-log54

Reviewer 4: Duplicated LOKI_BUCKET_NAME here. I suggest LOKI_BUCKET_NAME=s3-loki-bucket-example.

----

.Create s3 bucket
[source, terminal]
----
aws s3api create-bucket --bucket $LOKI_BUCKET_NAME --region $AWS_DEFAULT_REGION
aws s3api create-bucket --bucket $LOKI_BUCKET_NAME --region $AWS_DEFAULT_REGION --create-bucket-configuration LocationConstraint=$AWS_DEFAULT_REGION
----

.Create Storage Secret
[source, terminal]
----
oc -n "${NAMESPACE}" delete secret ${SECRETNAME}
oc -n "${NAMESPACE}" create secret generic ${SECRETNAME} \
    --from-literal=endpoint=${ENDPOINT} \
    --from-literal=region=${REGION} \
    --from-literal=bucketnames=${LOKI_BUCKET_NAME} \
    --from-literal=AWS_ACCESS_KEY_ID=${AWS_ACCESS_KEY_ID} \
    --from-literal=access_key_secret=${AWS_SECRET_ACCESS_KEY}
----

.Deploy lokistack under openshift-logging namespace
[source, terminal]
----

> oc process -f aosqe-tools/logging/log_template/loki-operator/lokistack_s3_template.yaml  -p STORAGE_CLASS=${STORAGE_CLASS} -p SECRET_NAME=${SECRETNAME} |oc create -f -

Reviewer 4: it is private link 'aosqe-tools/logging/log_template/loki-operator/'. Let's show template directly.

```
cat <<EOF  | tee  lokistack_s3_example.yaml
{
    "apiVersion": "loki.grafana.com/v1beta1",
    "kind": "LokiStack",
    "metadata": {
        "name": "lokistack-sample"
    },
    "spec": {
        "managementState": "Managed",
        "replicationFactor": 1,
        "size": "1x.extra-small",
        "storage": {
            "secret": {
                "name": "credsecret",
                "type": "s3"

            }
        },
        "storageClassName": "gp2",
        "tenants": {
            "mode": "openshift-logging"
        }
    } 
}
EOF
oc create -f  lokistack_s3_example.yaml
```

> ----

Reviewer 3: I think we should be displaying the components that are created as a part of the mentioned LokiStack above.

$ oc get pods -n openshift-logging
NAME                                           READY   STATUS    RESTARTS   AGE
lokistack-dev-compactor-0                      1/1     Running   0          63m
lokistack-dev-distributor-8696c45889-slb9k     1/1     Running   0          63m
lokistack-dev-gateway-5bfc4bdd67-snjjl         2/2     Running   0          63m
lokistack-dev-index-gateway-0                  1/1     Running   0          63m
lokistack-dev-ingester-0                       1/1     Running   0          63m
lokistack-dev-querier-9ccd98f9d-f6c5k          1/1     Running   0          63m
lokistack-dev-query-frontend-58bc479b7-dp8h9   1/1     Running   0          63m
